<a href="https://colab.research.google.com/github/Locung201/-HCMUS-BI2425/blob/main/Group3_Data_Preparation_%2B_Data_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


<style>
    body {
        display: flex;
        justify-content: center; /* Căn giữa theo chiều ngang */
        align-items: center; /* Căn giữa theo chiều dọc */
        height: 100vh; /* Chiều cao khung hình */
        margin: 0;
    }
    table {
        width: 80%;
        margin: auto;
        border-collapse: collapse;
        font-size: 20px;
    }
    th, td {
        border: 1px solid #dddddd;
        text-align: center;
        padding: 15px;
    }
    th {
        background-color: #4CAF50;
        color: white;
        font-size: 22px;
    }
    tr:nth-child(even) {
        background-color: #f9f9f9;
    }
    tr:nth-child(odd) {
        background-color: #ffffff;
    }
    h2 {
        color: #4CAF50;
        text-align: center;
        font-size: 26px;
    }
    .icon {
        color: #4CAF50;
        font-size: 22px;
        margin-right: 8px;
    }
    a {
        text-decoration: none;
        color: #4CAF50;
    }
</style>

<h2>Nhóm 3 - CSC12110 - PTDLUD</h2>

<table>
    <tr>
        <th>MSSV</th>
        <th>Họ Tên</th>
        <th>Email</th>
    </tr>
    <tr>
        <td>18127008</td>
        <td><i class="icon">👨‍💻</i> Lê Mạnh Hoàng</td>
        <td><a href="mailto:lmhoang18@clc.fitus.edu.vn">lmhoang18@clc.fitus.edu.vn</a></td>
    </tr>
    <tr>
        <td>21127211</td>
        <td><i class="icon">👩‍💻</i> Nguyễn Vũ Tường An</td>
        <td><a href="mailto:nvtan21@clc.fitus.edu.vn">nvtan21@clc.fitus.edu.vn</a></td>
    </tr>
    <tr>
        <td>21127699</td>
        <td><i class="icon">👩‍💻</i> Lô Thủy Tiên</td>
        <td><a href="mailto:lttien21@clc.fitus.edu.vn">lttien21@clc.fitus.edu.vn</a></td>
    </tr>
</table>

# <span style="background-color: #D3D3D3; font-weight: bold;">Giai đoạn 1: Chuẩn bị Dữ liệu (Data Preparation) + Data Preprocessing</span>
## <span style="background-color: #D3D3D3; font-weight: bold;">1. Tải dữ liệu từ nguồn cung cấp</span>

In [6]:
# Uploading files from your local file system
from google.colab import files
uploaded = files.upload()

# Đọc file CSV đã upload
import pandas as pd

# Lấy tên file đã upload
filename = list(uploaded.keys())[0]

# Đọc dữ liệu từ file CSV
data = pd.read_csv(filename, index_col=0)

# Hiển thị vài dòng đầu tiên của dữ liệu
data.head()

Saving 2425_QC.csv to 2425_QC.csv


Age  Area Income  Daily Internet Usage  \
Daily Time Spent on Site                                            
68.95                     35.0     61833.90                256.09   
80.23                     31.0     68441.85                193.77   
69.47                     26.0     59785.94                236.50   
74.15                     29.0     54806.18                245.89   
68.37                     35.0     73889.99                225.58   

                                                  Ad Topic Line  \
Daily Time Spent on Site                                          
68.95                        Cloned 5thgeneration orchestration   
80.23                        Monitored national standardization   
69.47                          Organic bottom-line service-desk   
74.15                     Triple-buffered reciprocal time-frame   
68.37                             Robust logistical utilization   

                                    City  Male     Country        Timestamp  \
Daily Time Spent on Site                                                      
68.95                        Wrightburgh     0     Tunisia   27/3/2016 0:53   
80.23                          West Jodi     1       Nauru    4/4/2016 1:39   
69.47                           Davidton     0  San Marino  13/3/2016 20:35   
74.15                     West Terrifurt     1       Italy   10/1/2016 2:31   
68.37                       South Manuel     0     Iceland    3/6/2016 3:36   

                          Clicked on Ad  
Daily Time Spent on Site                 
68.95                                 0  
80.23                                 0  
69.47                                 0  
74.15                                 0  
68.37                                 0

###2. Kiểm tra cấu trúc dữ liệu

In [7]:
print("Number of rows and columns in the data:", data.shape)

print("Information about the data structure:")
data.info()

print("Names of the attributes in the dataset:", data.columns)

print("Some top rows of the data:")
data.head()

Number of rows and columns in the data: (1002, 9)
Information about the data structure:
<class 'pandas.core.frame.DataFrame'>
Index: 1002 entries, 68.95 to 36.91
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Age                   1001 non-null   float64
 1   Area Income           1002 non-null   float64
 2   Daily Internet Usage  1002 non-null   float64
 3   Ad Topic Line         1002 non-null   object 
 4   City                  1002 non-null   object 
 5   Male                  1002 non-null   int64  
 6   Country               1002 non-null   object 
 7   Timestamp             1002 non-null   object 
 8   Clicked on Ad         1002 non-null   int64  
dtypes: float64(3), int64(2), object(4)
memory usage: 78.3+ KB
Names of the attributes in the dataset: Index(['Age', 'Area Income', 'Daily Internet Usage', 'Ad Topic Line', 'City',
       'Male', 'Country', 'Timestamp', 'Clicked on Ad'],
      d

Age  Area Income  Daily Internet Usage  \
Daily Time Spent on Site                                            
68.95                     35.0     61833.90                256.09   
80.23                     31.0     68441.85                193.77   
69.47                     26.0     59785.94                236.50   
74.15                     29.0     54806.18                245.89   
68.37                     35.0     73889.99                225.58   

                                                  Ad Topic Line  \
Daily Time Spent on Site                                          
68.95                        Cloned 5thgeneration orchestration   
80.23                        Monitored national standardization   
69.47                          Organic bottom-line service-desk   
74.15                     Triple-buffered reciprocal time-frame   
68.37                             Robust logistical utilization   

                                    City  Male     Country        Timestamp  \
Daily Time Spent on Site                                                      
68.95                        Wrightburgh     0     Tunisia   27/3/2016 0:53   
80.23                          West Jodi     1       Nauru    4/4/2016 1:39   
69.47                           Davidton     0  San Marino  13/3/2016 20:35   
74.15                     West Terrifurt     1       Italy   10/1/2016 2:31   
68.37                       South Manuel     0     Iceland    3/6/2016 3:36   

                          Clicked on Ad  
Daily Time Spent on Site                 
68.95                                 0  
80.23                                 0  
69.47                                 0  
74.15                                 0  
68.37                                 0

<div style="text-align: center; color: #4CAF50; font-size: 20px;">
  <strong>Nhận xét về Dữ liệu:</strong>
</div>
- **Số dòng và cột**: Dữ liệu có 1002 dòng và 9 cột. Điều này cho thấy dữ liệu có số lượng mẫu khá lớn (1002 mẫu) với 9 thuộc tính.

- **Các cột trong dữ liệu**:
    1. **Age**: Độ tuổi của người dùng, có 1001 giá trị hợp lệ (1 giá trị thiếu). Dữ liệu có kiểu `float64`.
    2. **Area Income**: Thu nhập khu vực của người dùng, không có giá trị thiếu và có kiểu `float64`.
    3. **Daily Internet Usage**: Sử dụng internet hàng ngày của người dùng, không có giá trị thiếu và có kiểu `float64`.
    4. **Ad Topic Line**: Dòng quảng cáo mà người dùng đã tương tác, không có giá trị thiếu và có kiểu `object` (chuỗi).
    5. **City**: Thành phố của người dùng, không có giá trị thiếu và có kiểu `object` (chuỗi).
    6. **Male**: Giới tính của người dùng (0 cho nữ, 1 cho nam), không có giá trị thiếu và có kiểu `int64`.
    7. **Country**: Quốc gia của người dùng, không có giá trị thiếu và có kiểu `object` (chuỗi).
    8. **Timestamp**: Thời gian người dùng tương tác với quảng cáo, không có giá trị thiếu và có kiểu `object` (chuỗi). Cần chuyển đổi sang kiểu dữ liệu thời gian (`datetime`) để dễ dàng xử lý và phân tích.
    9. **Clicked on Ad**: Biến mục tiêu cho bài toán phân loại, cho biết người dùng có nhấp vào quảng cáo hay không (0 hoặc 1), không có giá trị thiếu và có kiểu `int64`.

- **Nhận xét**:
    - Dữ liệu có đầy đủ thông tin cần thiết, với một số giá trị thiếu trong cột `Age` (1 giá trị thiếu).
    - Các cột có kiểu dữ liệu hợp lý, nhưng cột `Timestamp` cần chuyển đổi thành kiểu `datetime` để thuận tiện cho việc phân tích theo thời gian.
    - Các thuộc tính như độ tuổi, thu nhập khu vực, và mức độ sử dụng internet có thể đóng vai trò quan trọng trong việc phân tích.
    - Biến mục tiêu `Clicked on Ad` là một biến nhị phân, và dữ liệu này sẽ được sử dụng để xây dựng mô hình phân loại.


# <span style="background-color: #D3D3D3; font-weight: bold;">II.Khám phá và tiền xử lý dữ liệu (Data Preprocessing)</span>


## <span style="background-color: #D3D3D3; font-weight: bold;">1. Handling Missing Value & Empty Data:</span>


Mô tả số lượng missing value của mỗi thuộc tính

In [ ]:
# Kiểm tra các giá trị bị thiếu
print("\nMissing Values:")
missing_values = data.isnull().sum()
print(missing_values)


Missing Values:
Age                     1
Area Income             0
Daily Internet Usage    0
Ad Topic Line           0
City                    0
Male                    0
Country                 0
Timestamp               0
Clicked on Ad           0
dtype: int64


<div style="text-align: center; color: #4CAF50; font-size: 20px;">
  <strong>Nhận xét:</strong>
</div>

<p style="font-size: 16px; text-align: center; color: #555555;">
Chỉ có 1 giá trị bị thiếu trong cột Age.
</p>

 Xử lý dữ liệu missing : Việc điền giá trị thiếu trong cột "Age" bằng giá trị trung vị giúp đảm bảo rằng dữ liệu không bị thiên lệch do giá trị ngoại lệ và vẫn duy trì tính chính xác cho các phân tích và mô hình dự đoán tiếp theo.


In [ ]:
# Điền giá trị thiếu cho cột 'Age' bằng giá trị trung vị
data['Age'].fillna(data['Age'].median(), inplace=True)

# Kiểm tra lại số lượng giá trị thiếu sau khi điền
print("\nMissing Values After Filling:")
missing_values_after = data.isnull().sum()
print(missing_values_after)


Missing Values After Filling:
Age                     0
Area Income             0
Daily Internet Usage    0
Ad Topic Line           0
City                    0
Male                    0
Country                 0
Timestamp               0
Clicked on Ad           0
dtype: int64


<ipython-input-22-4bac8d6e0cb4>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].median(), inplace=True)


<div style="text-align: center; color: #4CAF50; font-size: 20px;">
  <strong>Nhận xét:</strong>
</div>

<p style="font-size: 16px; text-align: center; color: #555555;">
Kết quả kiểm tra cho thấy không còn giá trị nào bị thiếu trong tất cả các cột, điều này cho thấy dữ liệu hiện tại đã đầy đủ và sẵn sàng cho các phân tích tiếp theo.
</p>



## <span style="background-color: #D3D3D3; font-weight: bold;">2.Handling Duplications | Errors:</span>


Mô tả số lượng Duplications | Errors

In [ ]:
# Kiểm tra các dòng trùng lặp
duplicates = data.duplicated()
print("\nDuplicate rows:")
print(duplicates)


Duplicate rows:
Daily Time Spent on Site
68.95    False
80.23    False
69.47    False
74.15    False
68.37    False
         ...  
51.30    False
51.63    False
55.55    False
45.01    False
36.91     True
Length: 1002, dtype: bool


<div style="text-align: center; color: #4CAF50; font-size: 20px;">
  <strong>Nhận xét:</strong>
</div>

<p style="font-size: 16px; text-align: center; color: #555555;">
  Dữ liệu cho thấy rằng dòng có giá trị "Daily Time Spent on Site" là <strong>36.91</strong> có bản sao trùng lặp, được đánh dấu là <strong>True</strong>, trong khi các dòng khác đều được đánh dấu là <strong>False</strong>, cho thấy chúng là các giá trị duy nhất.
</p>


Sử dụng phương thức drop_duplicates(keep='first') để giữ lại dòng đầu tiên của các dòng trùng lặp và xóa các dòng còn lại.

In [ ]:
# Giữ lại dòng đầu tiên và xóa các dòng trùng lặp
data_cleaned_first = data.drop_duplicates(keep='first')

# Hiển thị thông tin về DataFrame sau khi xóa các dòng trùng lặp
print("\nDataFrame after keeping the first occurrence:")
print(data_cleaned_first)

# Kiểm tra lại số lượng dòng trong DataFrame trước và sau khi xóa
print("\nNumber of rows before removal:", len(data))
print("Number of rows after removal:", len(data_cleaned_first))


DataFrame after keeping the first occurrence:
                           Age  Area Income  Daily Internet Usage  \
Daily Time Spent on Site                                            
68.95                     35.0     61833.90                256.09   
80.23                     31.0     68441.85                193.77   
69.47                     26.0     59785.94                236.50   
74.15                     29.0     54806.18                245.89   
68.37                     35.0     73889.99                225.58   
...                        ...          ...                   ...   
72.97                     30.0     71384.57                208.58   
51.30                     45.0     67782.17                134.42   
51.63                     51.0     42415.72                120.37   
55.55                     19.0     41920.79                187.95   
45.01                     35.0     29875.80                178.35   

                                                  Ad To

<div style="text-align: center; color: #4CAF50; font-size: 20px;">
  <strong>Nhận xét:</strong>
</div>

<p style="font-size: 16px; text-align: center; color: #555555;">
  Sau khi loại bỏ các giá trị trùng lặp trong DataFrame, số lượng dòng đã giảm từ 1002 xuống còn 1000, tức là có 2 dòng bị loại bỏ do trùng lặp.
</p>

<p style="font-size: 16px; text-align: center; color: #555555;">
  Điều này cho thấy rằng trong dữ liệu ban đầu có một số bản ghi trùng lặp, và việc loại bỏ các bản sao giúp dữ liệu trở nên chính xác và không bị ảnh hưởng bởi các dòng thừa. Điều này sẽ giúp đảm bảo rằng các phân tích và mô hình dự đoán tiếp theo không bị sai lệch hoặc có sự lặp lại không cần thiết.
</p>
